### DISCUSSION TITLE
Practical differences between standalone replication and distributed topology

### DISCUSSION THREAD

**Discussion by Agalin** [cloudnative-pg/cloudnative-pg#6326] [General]
## Background

Hi, I'm currently testing cloudnative-pg as an alternative to the currently used operator and trying to understand practical differences between the distributed topology (DT) and standalone replicas (SR). I'm aware that DT is a new feature and functionality may be extended in the future but it's strictly about current situation.

According to docs (differences between SR and DT):
> *  Lack of Distributed Database Concept: Standalone Replica Clusters do not support the concept of a distributed database, whether in simple forms (two clusters) or more complex configurations (e.g., three clusters in a circular topology).
> * No Global Primary Cluster: There is no notion of a global primary cluster in Standalone Replica Clusters.
> * No Controlled Switchover: A Standalone Replica Cluster can only be promoted to primary. The former primary cluster must be re-cloned, as controlled switchover is not possible.

There is also the architecture diagram:
![architecture diagram](https://cloudnative-pg.io/documentation/1.24/images/multi-cluster.png)

## Analysis

* SR lacks DT concepts... but what does it really mean? I don't see any real differences besides a slightly different way of defining if a specific cluster should replicate or be a primary.
* Global primary cluster, while present on the architecture diagram doesn't seem to really matter. PGBouncer defined for the replica cluster points to the replica primary in the RW mode - exactly as marked on the diagram. Replica cluster is fully read-only, I'm unaware of any magic that would make it writable.
* Controlled Switchover - to my understanding the only real advantage of using DT over SR. And the difference is that... we don't need to resync the old primary from scratch, right?

So... that's it? The whole ~9 pages long documentation about DT + manifest spec differs from the original replication method only in the fact you don't need to rebuild the old primary and only in case of a planned switchover of the whole cluster (which from my experience barely ever happens, likely only for major version upgrade through the logical replication which is not supported in this scenario anyway)?

## Conclusion and future

Am I missing something? If not then current state of the distributed topology is... disappointing. It doesn't really seem to help with fast failover compared to the standard replication flow. Slightly easier controlled cross-cluster switchover doesn't seem worth the effort.

'm aware of plans to allow adding DT clusters as quorum members. Hard to say for me how useful that feature may be. I'm mostly surprised there is no way to target the writeable primary through a PGBouncer deployment (the only thing lacking seems to be an option to specify a secret for it in the external clusters section, then proper config entry would be craftable). But to be honest, with basically everything being done manually during the failover it seems that the only real difference is in size of the kubectl patch to target another primary / switch to being a primary.

**Comment by gbartolini** [comment]
Thank you for raising this discussion and sharing your feedback. Let me address your points step by step to provide clarity.

### Cross-Cluster Failover in CloudNativePG

CloudNativePG does not support cross-cluster automated failover, as it is designed to operate within the boundaries of a single Kubernetes cluster and exploit availability zones for better RTO outcomes. Extending the operator's authority beyond this scope would significantly increase the project's complexity and deviate from our design philosophy of integrating seamlessly with the cloud-native ecosystem. 

This principle aligns with our application to the CNCF Sandbox, emphasising interoperability with Kubernetes-native tools rather than building a monolithic solution. In this context, we advocate for handling demotion and promotion declaratively, either through GitOps workflows or by delegating such operations to a higher-level authority, such as a multi-cluster manager or federation-aware system.

### Out of Scope by Design

Automated failover across multiple Kubernetes clusters (and by extension, regions) has never been within the scope of CloudNativePG. Adding such functionality would conflict with our current architecture and objectives, unnecessarily increasing the already high complexity of the project.

---

### Response to Your Specific Questions

> **"SR lacks DT concepts... but what does it really mean? I don't see any real differences besides a slightly different way of defining if a specific cluster should replicate or be a primary."**

The difference is more significant than it may seem. A standalone replica (SR) cluster does not have the concept of a primary cluster spanning multiple Kubernetes clusters. In contrast, a distributed topology (DT) introduces this concept, by adding a field for it. However, even DT it requires external coordination and control by a higher authority as it spans multiple clusters, but it makes switchover possible and declarative (something you might be underestimating).

This distinction is fundamental to understanding the scope and applicability of each approach.

> **"Controlled Switchover - to my understanding the only real advantage of using DT over SR. And the difference is that... we don't need to resync the old primary from scratch, right?"**

This feature has been a game-changer for users managing multi-terabyte databases, as they can avoid the overhead of resynchronising a large dataset from scratch. While your use case might not highlight the full benefits, for others, this capability significantly reduces downtime and operational burden, making it invaluable in production scenarios. This scenario has been appreciated in the financial sector, specifically.

---

### Closing Thoughts

I'm sorry if this feature does not align with your expectations. We recognise that cross-cluster failover is a challenging feature to implement, especially while maintaining the core principles of our project. CloudNativePG is open source, and contributions are always welcome. If you have a specific vision for how this feature could be implemented, we’d be happy to review a pull request and collaborate to explore its feasibility.

Thank you again for your thoughtful feedback—it’s through discussions like these that we can continue to improve CloudNativePG.

**Reply by Agalin** [reply]
Thanks for extensive answer. I appreciate faster replica restores after failover (managing some databases that are very busy causing long restores too). Still feels to me like a lot of config just for that, but how I fee about it is not important. 🙂 Good to know automatic failover is a non-goal.

Going back to the primary cluster, is there anything defined primary is used for _besides_ replication? While it's supposed to be managed externally I'd still expect it to be reflected in at least a dedicated service (targetting the external primary) and an option to create a pgbouncer targetting this service instead of the local primary. Is anything like that planned for the future?

**Reply by gbartolini** [reply]
No worries, thanks for asking! Let me understand, you are using this setup in the same region?

**Reply by Agalin** [reply]
Right now I'm evaluating CNPG for migration from Crunchy Data.

We have some services (in different regions) that are fine with read-only local replicas but still need access to primary at times.

Currently it's handled by some clever labelling together with linkerd service mirroring + trafficspllit to provide setup where each cluster has a service pointing to the current primary. This setup would also work with CNPG but it's quite complicated. Changing just the topology config to get such service would be way easier to maintain.
